zkevm-circuits/src/rlp_circuit_fsm.rs

```diff
@@ -1405,12 +1497,489 @@ impl<F: Field> RlpCircuitConfig<F> {
+            );
+            cb.require_boolean(
+                "is_stack_push is binary",
+                meta.query_advice(rlp_decoding_table.is_stack_push, Rotation::cur()),
+            );
+            cb.require_boolean(
+                "is_stack_pop is binary",
+                meta.query_advice(rlp_decoding_table.is_stack_pop, Rotation::cur()),
+            );
+            cb.require_boolean(
+                "is_stack_update is binary",
+                meta.query_advice(rlp_decoding_table.is_stack_update, Rotation::cur()),
+            );
+
+            // Stack Init
+            cb.condition(
+                meta.query_advice(rlp_decoding_table.is_stack_init, Rotation::cur()),
+                |cb| {
+                    cb.require_zero(
+                        "stack inits at depth 0",
+                        meta.query_advice(rlp_decoding_table.depth, Rotation::cur()),
+                    );
+                    cb.require_equal(
+                        "stack init pushes all remaining_bytes onto depth 0",
+                        meta.query_advice(rlp_decoding_table.value, Rotation::cur()),
+                        meta.query_advice(byte_rev_idx, Rotation::cur()),
+                    );
+                    cb.require_equal(
+                        "stack can only init once with the first byte",
+                        meta.query_advice(byte_idx, Rotation::cur()),
+                        1.expr(),
+                    );
+                },
+            );
+
+            // Stack depth must stay same or only increment by 1 when the stack is not initiating
+            cb.condition(
+                not::expr(meta.query_advice(rlp_decoding_table.is_stack_init, Rotation::cur())),
+                |cb| {
+                    cb.require_zero(
+                        "stack depth stays the same or increments by 1",
+                        (meta.query_advice(rlp_decoding_table.depth, Rotation::cur())
+                            - meta.query_advice(rlp_decoding_table.depth, Rotation::prev()))
+                            * (meta.query_advice(rlp_decoding_table.depth, Rotation::cur())
+                                - meta.query_advice(rlp_decoding_table.depth, Rotation::prev())
+                                - 1.expr()),
+                    );
+                },
+            );
+
+            // Stack Push
+            cb.condition(
+                meta.query_advice(rlp_decoding_table.is_stack_push, Rotation::cur()),
+                |cb| {
+                    cb.require_zero(
+                        "The higher depth must have 0 bytes.",
+                        meta.query_advice(rlp_decoding_table.value_prev, Rotation::cur()),
+                    );
+                    cb.require_zero(
+                        "Before a new PUSH, all previous bytes must be processed.",
+                        meta.query_advice(rlp_decoding_table.value, Rotation::prev()),
+                    );
```

Comment:
As an example, assume a `PUSH` operation pushes bytes onto depth 3. 

=> By the stack ops sorting rules, all depth 2 stack ops should precede this `PUSH` op. 
=> For any stack depth, all bytes must be processed (no bytes were skipped).
=> Because all bytes on a depth must be processed and all the stack ops processing depth 2 bytes are already listed before the `PUSH` op onto depth 3, therefore, the `value` on the previous row (for a stack op in depth 2) must be 0.